primitives/treasury-funding/src/lib.rs

```diff
@@ -0,0 +1,396 @@
+        let system_utilisation =
+            P::from_rational(used_amount.into().into(), provided_amount.into().into());
+        let ideal_system_utilisation = T::IdealUtilisationRate::get();
+        let falloff = T::DecayRate::get();
+
+        // Calculate the adjustment to be used to calculate the treasury cut.
+        let adjustment = compute_adjustment_over_minimum_cut(
+            system_utilisation,
+            ideal_system_utilisation,
+            falloff,
+        );
+
+        // Get the minimum and maximum cut from the configuration and calculate the difference between them.
+        let minimum_cut = T::MinimumCut::get();
+        let maximum_cut = T::MaximumCut::get();
+        let delta_cut = maximum_cut.saturating_sub(minimum_cut);
+
+        // Calculate the final treasury cut percentage by adjusting the minimum cut with the adjustment times the delta cut.
+        let treasury_cut = minimum_cut.saturating_add(delta_cut * adjustment);
+
+        // Calculate the amount to transfer to the treasury by using the equation `amount_to_charge * treasury_cut`,
+        // where `treasury_cut` is the percentage of funds that should go to the treasury. `mul_floor` is used to
+        // round down the result to the nearest integer.
+        treasury_cut.mul_floor::<Self::Balance>(amount_to_charge)
+    }
+}
+
+/// Compute the adjustment percentage of the treasury cut, which is the percentage of the delta between the minimum
+/// and maximum treasury cut that should be added to the minimum cut. The adjustment is calculated using the following formula:
+///
+/// ```ignore
+/// I(x) = for x between 0 and x_ideal: 1 - x / x_ideal,
+/// for x between x_ideal and 1: 1 - 2^((x_ideal - x) / d)
+/// ```
+///
+/// where:
+/// * x is the system utilisation rate, i.e. fraction of the total storage being provided in the system
+/// that is currently being used.
+/// * d is the falloff or `decay_rate`
+/// * x_ideal: the ideal system utilisation rate.
+///
+/// The result is meant to be scaled with the minimum percentage amount and maximum percentage amount
+///  that goes to the treasury.
+///
+/// Arguments are:
+/// * `system_utilisation`: The fraction of the total storage being provided in the system that is
+/// currently being used. Known as `x` in the literature. Must be between 0 and 1.
+/// * `ideal_system_utilisation`: The fraction of total storage being provided in the system that should
+/// be being actively used. Known as `x_ideal` in the literature. Must be between 0 and 1.
+/// * `falloff`: Known as `decay_rate` in the literature. A co-efficient dictating the strength of
+///   the global incentivization to get the `ideal_system_utilisation`. A higher number results in less typical
+///   funds to the treasury at the cost of greater volatility for providers. Must be more than 0.01.
+pub fn compute_adjustment_over_minimum_cut<P: PerThing>(
```

Comment:
We should reference that this calculation is heavily inspired in Polkadot's staking rate formula code.